sys/net/link_layer/ng_simplemac/ng_simplemac.c

```diff
@@ -0,0 +1,364 @@
+
+        if (sendto == NULL) {
+            DEBUG("simplemac: unable to forward packet of type %i\n",
+                  pkt->next->type);
+        }
+
+#endif
+
+        /* send the packet to everyone interested in it's type */
+        while (sendto != NULL) {
+            DEBUG("simplemac: sending pkt %p to PID %u\n", pkt, sendto->pid);
+            ng_netapi_send(sendto->pid, pkt);
+            sendto = ng_netreg_getnext(sendto);
+        }
+    }
+}
+
+/**
+ * @brief     wait function that sets timer event. Wait time depends on backoff_exp.
+ *
+ * @param[in] backoff_exponent    wait-time depends on this exponent.
+ *
+ * @return number of bytes that were actually send out
+ * @return CHANNEL_ACCESS_FAILURE if packet could not sent out
+ */
+static int backoff_wait(uint8_t backoff_exponent)
+{
+    uint16_t random = 0;
+    uint16_t backoff_intervall = 0;
+    char buf[2];
+
+    if (backoff_exponent < MAX_BE) {
+        backoff_exponent = MAX_BE;
+    }
+
+    /* backoff_intervall is defined as ((2^bf)-1)*SYMBOL_LENGTH */
+    backoff_intervall = 1;
+    backoff_intervall <<= backoff_exponent;
+    backoff_intervall--;
+    backoff_intervall *= SYMBOL_LENGTH;
+    DEBUG("upper limit for backoff_intervall: %i\n", backoff_intervall);
+
+    /* Generate 16bit random number */
+    if (random_read(buf, 2) != 2) {
+        return 0;   /* what to return in case of random_read error? */
+    }
+
+    random = buf[0];
+    random <<= 8;
+    random = random | (0x00ff & buf[1]);
+
+    backoff_intervall = random % backoff_intervall;
+    DEBUG("Random number generated: %i\n", random);
+    DEBUG("Wait interval: %i\n", backoff_intervall);
+
+    timer_set(mac_tmr, MAC_TMR_CH, backoff_intervall);
+    DEBUG("timer set to %i\n", random);
+    return 0;
+}
+
+/**
+ * @brief     The mac_send_statechart function implements an statechart for better clarity
+ *
+ * @param[in] dev   network device descriptor
+ * @param[in] pkt   pointer to the data in the packet buffer
+ *
+ * @return number of bytes that were actually send out
+ * @return -1 if packet could not sent out
+ */
+static void _mac_send_statechart()
+{
+    ng_netapi_opt_t *conf;
+    int res;
+
+    while (1) {
+        switch (simplemac_state) {
+
+            case CSMA_IDLE:
+                DEBUG("simplemac-state: CSMA_IDLE");
+                be = 0;
+                retries = 0;
+                simplemac_state = CSMA_PERFORM_CCA;
+                break;
+
+            case CSMA_PERFORM_CCA:
+                DEBUG("simplemac-state: CSMA_PERFORM_CCA");
+                timer_clear(mac_tmr, MAC_TMR_CH);
+
+                if (be > MAX_CSMA_BACKOFFS) {
+                    return;
+                    simplemac_state = CSMA_IDLE;
+                }
+
+                conf -> opt = NETCONF_OPT_IS_CHANNEL_CLR;
+                conf -> data = 0;
+                /* TODO: What has to be assigned to data? */
+                /* In my opinion the CCA-check must be interrupt driven by the radio.
```

Comment:
Ok, what about to let the driver decide if the get(CCA) function is blocking or not? If the radio does not support interrupt for CCA, the function could be blocking and directly returning the CCA result.

If the radio supports interrupt for CCA or it has automatic CCA before transmission, the get(CCA) function would not be blocking but signalizing that there is an interrupt that has to be waited for, before executing tx.